coap_io.c: Cleanup MinGW builds

mrdeep1 opened the pull request:
Move the MinGW #if tests to after where mswsock.h is included.

Tidy up documentation.

Tested against MSYS2 UCRT64, CLANG64, MINGW32 and MINGW64 on Windows 11.
Fixed a couple of minor build issues found.

mrdeep1 committed:
coap_io.c: Cleanup MinGW builds

Move the MinGW #if tests to after where mswsock.h is included.

Tidy up documentation.

Tested against MSYS2 UCRT64, CLANG64, MINGW32 and MINGW64 on Windows 11.
Fixed a couple of minor build issues found.

mrdeep1 commented:
@jcelerier @anyc does this fix work for you?

anyc commented:
No, the build fails with i686-w64-mingw32 (13.2.0) on Ubuntu 24.04:

```
/home/libcoap/src/coap_io.c: In function ‘coap_socket_send’:
/home/libcoap/src/coap_io.c:854:14: warning: implicit declaration of function ‘CMSG_SPACE’; did you mean ‘C1_SPACE’? [-Wimplicit-function-declaration]
  854 |     char buf[CMSG_SPACE(sizeof(struct in6_pktinfo))];
      |              ^~~~~~~~~~
      |              C1_SPACE
/home/libcoap/src/coap_io.c:854:14: warning: nested extern declaration of ‘CMSG_SPACE’ [-Wnested-externs]
[ 31%] Built target doc_doxygen
[ 33%] Building C object CMakeFiles/coap-3.dir/src/coap_net.c.obj
/home/libcoap/src/coap_io.c:889:18: warning: implicit declaration of function ‘CMSG_FIRSTHDR’; did you mean ‘WSA_CMSG_FIRSTHDR’? [-Wimplicit-function-declaration]
  889 |           cmsg = CMSG_FIRSTHDR(&mhdr);
      |                  ^~~~~~~~~~~~~
      |                  WSA_CMSG_FIRSTHDR
/home/libcoap/src/coap_io.c:889:18: warning: nested extern declaration of ‘CMSG_FIRSTHDR’ [-Wnested-externs]
/home/libcoap/src/coap_io.c:889:16: warning: assignment to ‘struct cmsghdr *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
  889 |           cmsg = CMSG_FIRSTHDR(&mhdr);
      |                ^
/home/libcoap/src/coap_io.c:890:15: error: invalid use of undefined type ‘struct cmsghdr’
  890 |           cmsg->cmsg_level = COAP_SOL_IP;
      |               ^~
/home/libcoap/src/coap_io.c:891:15: error: invalid use of undefined type ‘struct cmsghdr’
  891 |           cmsg->cmsg_type = IP_PKTINFO;
      |               ^~
/home/libcoap/src/coap_io.c:892:15: error: invalid use of undefined type ‘struct cmsghdr’
  892 |           cmsg->cmsg_len = CMSG_LEN(sizeof(struct in_pktinfo));
      |               ^~
/home/libcoap/src/coap_io.c:892:28: warning: implicit declaration of function ‘CMSG_LEN’ [-Wimplicit-function-declaration]
  892 |           cmsg->cmsg_len = CMSG_LEN(sizeof(struct in_pktinfo));
      |                            ^~~~~~~~
/home/libcoap/src/coap_io.c:892:28: warning: nested extern declaration of ‘CMSG_LEN’ [-Wnested-externs]
/home/libcoap/src/coap_io.c:919:16: warning: assignment to ‘struct cmsghdr *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
  919 |           cmsg = CMSG_FIRSTHDR(&mhdr);
      |                ^
/home/libcoap/src/coap_io.c:920:15: error: invalid use of undefined type ‘struct cmsghdr’
  920 |           cmsg->cmsg_level = IPPROTO_IPV6;
      |               ^~
/home/libcoap/src/coap_io.c:921:15: error: invalid use of undefined type ‘struct cmsghdr’
  921 |           cmsg->cmsg_type = IPV6_PKTINFO;
      |               ^~
/home/libcoap/src/coap_io.c:922:15: error: invalid use of undefined type ‘struct cmsghdr’
  922 |           cmsg->cmsg_len = CMSG_LEN(sizeof(struct in6_pktinfo));
      |               ^~
/home/libcoap/src/coap_io.c:949:14: warning: assignment to ‘struct cmsghdr *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
  949 |         cmsg = CMSG_FIRSTHDR(&mhdr);
      |              ^
/home/libcoap/src/coap_io.c:950:13: error: invalid use of undefined type ‘struct cmsghdr’
  950 |         cmsg->cmsg_level = COAP_SOL_IP;
      |             ^~
/home/libcoap/src/coap_io.c:951:13: error: invalid use of undefined type ‘struct cmsghdr’
  951 |         cmsg->cmsg_type = IP_PKTINFO;
      |             ^~
/home/libcoap/src/coap_io.c:952:13: error: invalid use of undefined type ‘struct cmsghdr’
  952 |         cmsg->cmsg_len = CMSG_LEN(sizeof(struct in_pktinfo));
      |             ^~
/home/libcoap/src/coap_io.c: In function ‘coap_socket_recv’:
/home/libcoap/src/coap_io.c:1105:9: error: invalid use of undefined type ‘struct cmsghdr’
 1105 |     cmsg->cmsg_len = CMSG_LEN(sizeof(buf));
      |         ^~
/home/libcoap/src/coap_io.c:1106:9: error: invalid use of undefined type ‘struct cmsghdr’
 1106 |     cmsg->cmsg_level = -1;
      |         ^~
/home/libcoap/src/coap_io.c:1107:9: error: invalid use of undefined type ‘struct cmsghdr’
 1107 |     cmsg->cmsg_type = -1;
      |         ^~
/home/libcoap/src/coap_io.c:1160:17: warning: assignment to ‘struct cmsghdr *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
 1160 |       for (cmsg = CMSG_FIRSTHDR(&mhdr); cmsg; cmsg = CMSG_NXTHDR(&mhdr, cmsg)) {
      |                 ^
/home/libcoap/src/coap_io.c:1160:54: warning: implicit declaration of function ‘CMSG_NXTHDR’; did you mean ‘WSA_CMSG_NXTHDR’? [-Wimplicit-function-declaration]
 1160 |       for (cmsg = CMSG_FIRSTHDR(&mhdr); cmsg; cmsg = CMSG_NXTHDR(&mhdr, cmsg)) {
      |                                                      ^~~~~~~~~~~
      |                                                      WSA_CMSG_NXTHDR
/home/libcoap/src/coap_io.c:1160:54: warning: nested extern declaration of ‘CMSG_NXTHDR’ [-Wnested-externs]
/home/libcoap/src/coap_io.c:1160:52: warning: assignment to ‘struct cmsghdr *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
 1160 |       for (cmsg = CMSG_FIRSTHDR(&mhdr); cmsg; cmsg = CMSG_NXTHDR(&mhdr, cmsg)) {
      |                                                    ^
/home/libcoap/src/coap_io.c:1164:17: error: invalid use of undefined type ‘struct cmsghdr’
 1164 |         if (cmsg->cmsg_level == IPPROTO_IPV6 && cmsg->cmsg_type == IPV6_PKTINFO) {
      |                 ^~
/home/libcoap/src/coap_io.c:1164:53: error: invalid use of undefined type ‘struct cmsghdr’
 1164 |         if (cmsg->cmsg_level == IPPROTO_IPV6 && cmsg->cmsg_type == IPV6_PKTINFO) {
      |                                                     ^~
/home/libcoap/src/coap_io.c:1181:17: error: invalid use of undefined type ‘struct cmsghdr’
 1181 |         if (cmsg->cmsg_level == COAP_SOL_IP && cmsg->cmsg_type == IP_PKTINFO) {
      |                 ^~
/home/libcoap/src/coap_io.c:1181:52: error: invalid use of undefined type ‘struct cmsghdr’
 1181 |         if (cmsg->cmsg_level == COAP_SOL_IP && cmsg->cmsg_type == IP_PKTINFO) {
      |                                                    ^~
/home/libcoap/src/coap_io.c:1206:17: error: invalid use of undefined type ‘struct cmsghdr’
 1206 |         if (cmsg->cmsg_level == IPPROTO_IP && cmsg->cmsg_type == IP_RECVDSTADDR) {
      |                 ^~
/home/libcoap/src/coap_io.c:1206:51: error: invalid use of undefined type ‘struct cmsghdr’
 1206 |         if (cmsg->cmsg_level == IPPROTO_IP && cmsg->cmsg_type == IP_RECVDSTADDR) {
      |                                                   ^~
/home/libcoap/src/coap_io.c:1218:19: error: invalid use of undefined type ‘struct cmsghdr’
 1218 |           if (cmsg->cmsg_level != -1 && cmsg->cmsg_type != -1) {
      |                   ^~
/home/libcoap/src/coap_io.c:1218:45: error: invalid use of undefined type ‘struct cmsghdr’
 1218 |           if (cmsg->cmsg_level != -1 && cmsg->cmsg_type != -1) {
      |                                             ^~
In file included from /home/libcoap/include/coap3/coap_net.h:39,
                 from /home/libcoap/include/coap3/coap_async.h:20,
                 from /home/libcoap/include/coap3/coap.h:47,
                 from /home/libcoap/include/coap3/coap_internal.h:83,
                 from /home/libcoap/include/coap3/coap_libcoap_build.h:45,
                 from /home/libcoap/src/coap_io.c:16:
/home/libcoap/src/coap_io.c:1220:32: error: invalid use of undefined type ‘struct cmsghdr’
 1220 |                            cmsg->cmsg_level, cmsg->cmsg_type);
      |                                ^~
/home/libcoap/include/coap3/coap_debug.h:286:30: note: in definition of macro ‘coap_log’
  286 |       coap_log_impl((level), __VA_ARGS__); \
      |                              ^~~~~~~~~~~
/home/libcoap/src/coap_io.c:1219:13: note: in expansion of macro ‘coap_log_debug’
 1219 |             coap_log_debug("cmsg_level = %d and cmsg_type = %d not supported - fix\n",
      |             ^~~~~~~~~~~~~~
/home/libcoap/src/coap_io.c:1220:50: error: invalid use of undefined type ‘struct cmsghdr’
 1220 |                            cmsg->cmsg_level, cmsg->cmsg_type);
      |                                                  ^~
/home/libcoap/include/coap3/coap_debug.h:286:30: note: in definition of macro ‘coap_log’
  286 |       coap_log_impl((level), __VA_ARGS__); \
      |                              ^~~~~~~~~~~
/home/libcoap/src/coap_io.c:1219:13: note: in expansion of macro ‘coap_log_debug’
 1219 |             coap_log_debug("cmsg_level = %d and cmsg_type = %d not supported - fix\n",
      |             ^~~~~~~~~~~~~~
make[2]: *** [CMakeFiles/coap-3.dir/build.make:227: CMakeFiles/coap-3.dir/src/coap_io.c.obj] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles/Makefile2:91: CMakeFiles/coap-3.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

mrdeep1 commented:
@anyc This should now work for you.

anyc commented:
Yes, it builds for i686 and x86_64 here.

mrdeep1 commented:
@anyc Thanks for confirming.

jcelerier commented:
can confirm that this fixes the issue on my end too, thanks!

Tested on MSYS2 clang64, ucrt64 and mingw64 on windows 11

jcelerier commented:
I am seeing hefty warnings on clang64 though, just in case: 

```

[31/183 21.0/sec] Building C object 3rdparty/libossia/3rdparty/libcoap/CMakeFiles/coap-3.dir/src/coap_io.c.obj
D:/dev/ossia/score/3rdparty/libossia/3rdparty/libcoap/src/coap_io.c:863:14: warning: defining a type within '__builtin_offsetof' is a C23 extension [-Wc23-extensions]
  863 |     char buf[CMSG_SPACE(sizeof(struct in6_pktinfo))];
      |              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
D:/dev/ossia/score/3rdparty/libossia/3rdparty/libcoap/src/coap_io.c:760:20: note: expanded from macro 'CMSG_SPACE'
  760 | #define CMSG_SPACE WSA_CMSG_SPACE
      |                    ^
D:/msys64/clang64/include/mswsock.h:189:73: note: expanded from macro 'WSA_CMSG_SPACE'
  189 | #define WSA_CMSG_SPACE(length) (WSA_CMSGDATA_ALIGN(sizeof(WSACMSGHDR) + WSA_CMSGHDR_ALIGN(length)))
      |                                                                         ^~~~~~~~~~~~~~~~~~~~~~~~~
D:/msys64/clang64/include/mswsock.h:184:48: note: expanded from macro 'WSA_CMSGHDR_ALIGN'
  184 | #define WSA_CMSGHDR_ALIGN(length) (((length) + TYPE_ALIGNMENT(WSACMSGHDR)-1) & (~(TYPE_ALIGNMENT(WSACMSGHDR)-1)))
      |                                                ^~~~~~~~~~~~~~~~~~~~~~~~~~
D:/msys64/clang64/include/winnt.h:136:40: note: expanded from macro 'TYPE_ALIGNMENT'
  136 | #define TYPE_ALIGNMENT(t) FIELD_OFFSET(struct { char x; t test; }, test)
      |                                        ^~~~~~
D:/msys64/clang64/include/winnt.h:687:62: note: expanded from macro 'FIELD_OFFSET'
  687 | #define FIELD_OFFSET(Type, Field) ((LONG) __builtin_offsetof(Type, Field))
      |                                                              ^~~~
D:/msys64/clang64/include/mswsock.h:185:39: note: expanded from macro 'WSA_CMSGDATA_ALIGN'
  185 | #define WSA_CMSGDATA_ALIGN(length) (((length) + MAX_NATURAL_ALIGNMENT-1) & (~(MAX_NATURAL_ALIGNMENT-1)))
      |                                       ^~~~~~
D:/dev/ossia/score/3rdparty/libossia/3rdparty/libcoap/src/coap_io.c:863:14: warning: defining a type within '__builtin_offsetof' is a C23 extension [-Wc23-extensions]
  863 |     char buf[CMSG_SPACE(sizeof(struct in6_pktinfo))];
      |              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
D:/dev/ossia/score/3rdparty/libossia/3rdparty/libcoap/src/coap_io.c:760:20: note: expanded from macro 'CMSG_SPACE'
  760 | #define CMSG_SPACE WSA_CMSG_SPACE
      |                    ^
D:/msys64/clang64/include/mswsock.h:189:73: note: expanded from macro 'WSA_CMSG_SPACE'
  189 | #define WSA_CMSG_SPACE(length) (WSA_CMSGDATA_ALIGN(sizeof(WSACMSGHDR) + WSA_CMSGHDR_ALIGN(length)))
      |                                                                         ^~~~~~~~~~~~~~~~~~~~~~~~~
D:/msys64/clang64/include/mswsock.h:184:83: note: expanded from macro 'WSA_CMSGHDR_ALIGN'
  184 | #define WSA_CMSGHDR_ALIGN(length) (((length) + TYPE_ALIGNMENT(WSACMSGHDR)-1) & (~(TYPE_ALIGNMENT(WSACMSGHDR)-1)))
      |                                                                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~
D:/msys64/clang64/include/winnt.h:136:40: note: expanded from macro 'TYPE_ALIGNMENT'
  136 | #define TYPE_ALIGNMENT(t) FIELD_OFFSET(struct { char x; t test; }, test)
      |                                        ^~~~~~
D:/msys64/clang64/include/winnt.h:687:62: note: expanded from macro 'FIELD_OFFSET'
  687 | #define FIELD_OFFSET(Type, Field) ((LONG) __builtin_offsetof(Type, Field))
      |                                                              ^~~~
D:/msys64/clang64/include/mswsock.h:185:39: note: expanded from macro 'WSA_CMSGDATA_ALIGN'
  185 | #define WSA_CMSGDATA_ALIGN(length) (((length) + MAX_NATURAL_ALIGNMENT-1) & (~(MAX_NATURAL_ALIGNMENT-1)))
      |                                       ^~~~~~
D:/dev/ossia/score/3rdparty/libossia/3rdparty/libcoap/src/coap_io.c:896:33: warning: defining a type within '__builtin_offsetof' is a C23 extension [-Wc23-extensions]
  896 |           mhdr.msg_controllen = CMSG_SPACE(sizeof(struct in_pktinfo));
      |                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
D:/dev/ossia/score/3rdparty/libossia/3rdparty/libcoap/src/coap_io.c:760:20: note: expanded from macro 'CMSG_SPACE'
  760 | #define CMSG_SPACE WSA_CMSG_SPACE
      |                    ^
D:/msys64/clang64/include/mswsock.h:189:73: note: expanded from macro 'WSA_CMSG_SPACE'
  189 | #define WSA_CMSG_SPACE(length) (WSA_CMSGDATA_ALIGN(sizeof(WSACMSGHDR) + WSA_CMSGHDR_ALIGN(length)))
      |                                                                         ^~~~~~~~~~~~~~~~~~~~~~~~~
D:/msys64/clang64/include/mswsock.h:184:48: note: expanded from macro 'WSA_CMSGHDR_ALIGN'
  184 | #define WSA_CMSGHDR_ALIGN(length) (((length) + TYPE_ALIGNMENT(WSACMSGHDR)-1) & (~(TYPE_ALIGNMENT(WSACMSGHDR)-1)))
      |                                                ^~~~~~~~~~~~~~~~~~~~~~~~~~
D:/msys64/clang64/include/winnt.h:136:40: note: expanded from macro 'TYPE_ALIGNMENT'
  136 | #define TYPE_ALIGNMENT(t) FIELD_OFFSET(struct { char x; t test; }, test)
      |                                        ^~~~~~
D:/msys64/clang64/include/winnt.h:687:62: note: expanded from macro 'FIELD_OFFSET'
  687 | #define FIELD_OFFSET(Type, Field) ((LONG) __builtin_offsetof(Type, Field))
      |                                                              ^~~~
D:/msys64/clang64/include/mswsock.h:185:39: note: expanded from macro 'WSA_CMSGDATA_ALIGN'
  185 | #define WSA_CMSGDATA_ALIGN(length) (((length) + MAX_NATURAL_ALIGNMENT-1) & (~(MAX_NATURAL_ALIGNMENT-1)))
      |                                       ^~~~~~
D:/dev/ossia/score/3rdparty/libossia/3rdparty/libcoap/src/coap_io.c:896:33: warning: defining a type within '__builtin_offsetof' is a C23 extension [-Wc23-extensions]
  896 |           mhdr.msg_controllen = CMSG_SPACE(sizeof(struct in_pktinfo));
      |                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
D:/dev/ossia/score/3rdparty/libossia/3rdparty/libcoap/src/coap_io.c:760:20: note: expanded from macro 'CMSG_SPACE'
  760 | #define CMSG_SPACE WSA_CMSG_SPACE
      |                    ^
D:/msys64/clang64/include/mswsock.h:189:73: note: expanded from macro 'WSA_CMSG_SPACE'
  189 | #define WSA_CMSG_SPACE(length) (WSA_CMSGDATA_ALIGN(sizeof(WSACMSGHDR) + WSA_CMSGHDR_ALIGN(length)))
      |                                                                         ^~~~~~~~~~~~~~~~~~~~~~~~~
D:/msys64/clang64/include/mswsock.h:184:83: note: expanded from macro 'WSA_CMSGHDR_ALIGN'
  184 | #define WSA_CMSGHDR_ALIGN(length) (((length) + TYPE_ALIGNMENT(WSACMSGHDR)-1) & (~(TYPE_ALIGNMENT(WSACMSGHDR)-1)))
      |                                                                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~
D:/msys64/clang64/include/winnt.h:136:40: note: expanded from macro 'TYPE_ALIGNMENT'
  136 | #define TYPE_ALIGNMENT(t) FIELD_OFFSET(struct { char x; t test; }, test)
      |                                        ^~~~~~
D:/msys64/clang64/include/winnt.h:687:62: note: expanded from macro 'FIELD_OFFSET'
  687 | #define FIELD_OFFSET(Type, Field) ((LONG) __builtin_offsetof(Type, Field))
      |                                                              ^~~~
D:/msys64/clang64/include/mswsock.h:185:39: note: expanded from macro 'WSA_CMSGDATA_ALIGN'
  185 | #define WSA_CMSGDATA_ALIGN(length) (((length) + MAX_NATURAL_ALIGNMENT-1) & (~(MAX_NATURAL_ALIGNMENT-1)))
      |                                       ^~~~~~
D:/dev/ossia/score/3rdparty/libossia/3rdparty/libcoap/src/coap_io.c:926:33: warning: defining a type within '__builtin_offsetof' is a C23 extension [-Wc23-extensions]
  926 |           mhdr.msg_controllen = CMSG_SPACE(sizeof(struct in6_pktinfo));
      |                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
D:/dev/ossia/score/3rdparty/libossia/3rdparty/libcoap/src/coap_io.c:760:20: note: expanded from macro 'CMSG_SPACE'
  760 | #define CMSG_SPACE WSA_CMSG_SPACE
      |                    ^
D:/msys64/clang64/include/mswsock.h:189:73: note: expanded from macro 'WSA_CMSG_SPACE'
  189 | #define WSA_CMSG_SPACE(length) (WSA_CMSGDATA_ALIGN(sizeof(WSACMSGHDR) + WSA_CMSGHDR_ALIGN(length)))
      |                                                                         ^~~~~~~~~~~~~~~~~~~~~~~~~
D:/msys64/clang64/include/mswsock.h:184:48: note: expanded from macro 'WSA_CMSGHDR_ALIGN'
  184 | #define WSA_CMSGHDR_ALIGN(length) (((length) + TYPE_ALIGNMENT(WSACMSGHDR)-1) & (~(TYPE_ALIGNMENT(WSACMSGHDR)-1)))
      |                                                ^~~~~~~~~~~~~~~~~~~~~~~~~~
D:/msys64/clang64/include/winnt.h:136:40: note: expanded from macro 'TYPE_ALIGNMENT'
  136 | #define TYPE_ALIGNMENT(t) FIELD_OFFSET(struct { char x; t test; }, test)
      |                                        ^~~~~~
D:/msys64/clang64/include/winnt.h:687:62: note: expanded from macro 'FIELD_OFFSET'
  687 | #define FIELD_OFFSET(Type, Field) ((LONG) __builtin_offsetof(Type, Field))
      |                                                              ^~~~
D:/msys64/clang64/include/mswsock.h:185:39: note: expanded from macro 'WSA_CMSGDATA_ALIGN'
  185 | #define WSA_CMSGDATA_ALIGN(length) (((length) + MAX_NATURAL_ALIGNMENT-1) & (~(MAX_NATURAL_ALIGNMENT-1)))
      |                                       ^~~~~~
D:/dev/ossia/score/3rdparty/libossia/3rdparty/libcoap/src/coap_io.c:926:33: warning: defining a type within '__builtin_offsetof' is a C23 extension [-Wc23-extensions]
  926 |           mhdr.msg_controllen = CMSG_SPACE(sizeof(struct in6_pktinfo));
      |                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
D:/dev/ossia/score/3rdparty/libossia/3rdparty/libcoap/src/coap_io.c:760:20: note: expanded from macro 'CMSG_SPACE'
  760 | #define CMSG_SPACE WSA_CMSG_SPACE
      |                    ^
D:/msys64/clang64/include/mswsock.h:189:73: note: expanded from macro 'WSA_CMSG_SPACE'
  189 | #define WSA_CMSG_SPACE(length) (WSA_CMSGDATA_ALIGN(sizeof(WSACMSGHDR) + WSA_CMSGHDR_ALIGN(length)))
      |                                                                         ^~~~~~~~~~~~~~~~~~~~~~~~~
D:/msys64/clang64/include/mswsock.h:184:83: note: expanded from macro 'WSA_CMSGHDR_ALIGN'
  184 | #define WSA_CMSGHDR_ALIGN(length) (((length) + TYPE_ALIGNMENT(WSACMSGHDR)-1) & (~(TYPE_ALIGNMENT(WSACMSGHDR)-1)))
      |                                                                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~
D:/msys64/clang64/include/winnt.h:136:40: note: expanded from macro 'TYPE_ALIGNMENT'
  136 | #define TYPE_ALIGNMENT(t) FIELD_OFFSET(struct { char x; t test; }, test)
      |                                        ^~~~~~
D:/msys64/clang64/include/winnt.h:687:62: note: expanded from macro 'FIELD_OFFSET'
  687 | #define FIELD_OFFSET(Type, Field) ((LONG) __builtin_offsetof(Type, Field))
      |                                                              ^~~~
D:/msys64/clang64/include/mswsock.h:185:39: note: expanded from macro 'WSA_CMSGDATA_ALIGN'
  185 | #define WSA_CMSGDATA_ALIGN(length) (((length) + MAX_NATURAL_ALIGNMENT-1) & (~(MAX_NATURAL_ALIGNMENT-1)))
      |                                       ^~~~~~
D:/dev/ossia/score/3rdparty/libossia/3rdparty/libcoap/src/coap_io.c:956:31: warning: defining a type within '__builtin_offsetof' is a C23 extension [-Wc23-extensions]
  956 |         mhdr.msg_controllen = CMSG_SPACE(sizeof(struct in_pktinfo));
      |                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
D:/dev/ossia/score/3rdparty/libossia/3rdparty/libcoap/src/coap_io.c:760:20: note: expanded from macro 'CMSG_SPACE'
  760 | #define CMSG_SPACE WSA_CMSG_SPACE
      |                    ^
D:/msys64/clang64/include/mswsock.h:189:73: note: expanded from macro 'WSA_CMSG_SPACE'
  189 | #define WSA_CMSG_SPACE(length) (WSA_CMSGDATA_ALIGN(sizeof(WSACMSGHDR) + WSA_CMSGHDR_ALIGN(length)))
      |                                                                         ^~~~~~~~~~~~~~~~~~~~~~~~~
D:/msys64/clang64/include/mswsock.h:184:48: note: expanded from macro 'WSA_CMSGHDR_ALIGN'
  184 | #define WSA_CMSGHDR_ALIGN(length) (((length) + TYPE_ALIGNMENT(WSACMSGHDR)-1) & (~(TYPE_ALIGNMENT(WSACMSGHDR)-1)))
      |                                                ^~~~~~~~~~~~~~~~~~~~~~~~~~
D:/msys64/clang64/include/winnt.h:136:40: note: expanded from macro 'TYPE_ALIGNMENT'
  136 | #define TYPE_ALIGNMENT(t) FIELD_OFFSET(struct { char x; t test; }, test)
      |                                        ^~~~~~
D:/msys64/clang64/include/winnt.h:687:62: note: expanded from macro 'FIELD_OFFSET'
  687 | #define FIELD_OFFSET(Type, Field) ((LONG) __builtin_offsetof(Type, Field))
      |                                                              ^~~~
D:/msys64/clang64/include/mswsock.h:185:39: note: expanded from macro 'WSA_CMSGDATA_ALIGN'
  185 | #define WSA_CMSGDATA_ALIGN(length) (((length) + MAX_NATURAL_ALIGNMENT-1) & (~(MAX_NATURAL_ALIGNMENT-1)))
      |                                       ^~~~~~
D:/dev/ossia/score/3rdparty/libossia/3rdparty/libcoap/src/coap_io.c:956:31: warning: defining a type within '__builtin_offsetof' is a C23 extension [-Wc23-extensions]
  956 |         mhdr.msg_controllen = CMSG_SPACE(sizeof(struct in_pktinfo));
      |                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
D:/dev/ossia/score/3rdparty/libossia/3rdparty/libcoap/src/coap_io.c:760:20: note: expanded from macro 'CMSG_SPACE'
  760 | #define CMSG_SPACE WSA_CMSG_SPACE
      |                    ^
D:/msys64/clang64/include/mswsock.h:189:73: note: expanded from macro 'WSA_CMSG_SPACE'
  189 | #define WSA_CMSG_SPACE(length) (WSA_CMSGDATA_ALIGN(sizeof(WSACMSGHDR) + WSA_CMSGHDR_ALIGN(length)))
      |                                                                         ^~~~~~~~~~~~~~~~~~~~~~~~~
D:/msys64/clang64/include/mswsock.h:184:83: note: expanded from macro 'WSA_CMSGHDR_ALIGN'
  184 | #define WSA_CMSGHDR_ALIGN(length) (((length) + TYPE_ALIGNMENT(WSACMSGHDR)-1) & (~(TYPE_ALIGNMENT(WSACMSGHDR)-1)))
      |                                                                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~
D:/msys64/clang64/include/winnt.h:136:40: note: expanded from macro 'TYPE_ALIGNMENT'
  136 | #define TYPE_ALIGNMENT(t) FIELD_OFFSET(struct { char x; t test; }, test)
      |                                        ^~~~~~
D:/msys64/clang64/include/winnt.h:687:62: note: expanded from macro 'FIELD_OFFSET'
  687 | #define FIELD_OFFSET(Type, Field) ((LONG) __builtin_offsetof(Type, Field))
      |                                                              ^~~~
D:/msys64/clang64/include/mswsock.h:185:39: note: expanded from macro 'WSA_CMSGDATA_ALIGN'
  185 | #define WSA_CMSGDATA_ALIGN(length) (((length) + MAX_NATURAL_ALIGNMENT-1) & (~(MAX_NATURAL_ALIGNMENT-1)))
      |                                       ^~~~~~
D:/dev/ossia/score/3rdparty/libossia/3rdparty/libcoap/src/coap_io.c:1089:14: warning: defining a type within '__builtin_offsetof' is a C23 extension [-Wc23-extensions]
 1089 |     char buf[CMSG_SPACE(sizeof(struct in6_pktinfo))];
      |              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
D:/dev/ossia/score/3rdparty/libossia/3rdparty/libcoap/src/coap_io.c:760:20: note: expanded from macro 'CMSG_SPACE'
  760 | #define CMSG_SPACE WSA_CMSG_SPACE
      |                    ^
D:/msys64/clang64/include/mswsock.h:189:73: note: expanded from macro 'WSA_CMSG_SPACE'
  189 | #define WSA_CMSG_SPACE(length) (WSA_CMSGDATA_ALIGN(sizeof(WSACMSGHDR) + WSA_CMSGHDR_ALIGN(length)))
      |                                                                         ^~~~~~~~~~~~~~~~~~~~~~~~~
D:/msys64/clang64/include/mswsock.h:184:48: note: expanded from macro 'WSA_CMSGHDR_ALIGN'
  184 | #define WSA_CMSGHDR_ALIGN(length) (((length) + TYPE_ALIGNMENT(WSACMSGHDR)-1) & (~(TYPE_ALIGNMENT(WSACMSGHDR)-1)))
      |                                                ^~~~~~~~~~~~~~~~~~~~~~~~~~
D:/msys64/clang64/include/winnt.h:136:40: note: expanded from macro 'TYPE_ALIGNMENT'
  136 | #define TYPE_ALIGNMENT(t) FIELD_OFFSET(struct { char x; t test; }, test)
      |                                        ^~~~~~
D:/msys64/clang64/include/winnt.h:687:62: note: expanded from macro 'FIELD_OFFSET'
  687 | #define FIELD_OFFSET(Type, Field) ((LONG) __builtin_offsetof(Type, Field))
      |                                                              ^~~~
D:/msys64/clang64/include/mswsock.h:185:39: note: expanded from macro 'WSA_CMSGDATA_ALIGN'
  185 | #define WSA_CMSGDATA_ALIGN(length) (((length) + MAX_NATURAL_ALIGNMENT-1) & (~(MAX_NATURAL_ALIGNMENT-1)))
      |                                       ^~~~~~
D:/dev/ossia/score/3rdparty/libossia/3rdparty/libcoap/src/coap_io.c:1089:14: warning: defining a type within '__builtin_offsetof' is a C23 extension [-Wc23-extensions]
 1089 |     char buf[CMSG_SPACE(sizeof(struct in6_pktinfo))];
      |              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
D:/dev/ossia/score/3rdparty/libossia/3rdparty/libcoap/src/coap_io.c:760:20: note: expanded from macro 'CMSG_SPACE'
  760 | #define CMSG_SPACE WSA_CMSG_SPACE
      |                    ^
D:/msys64/clang64/include/mswsock.h:189:73: note: expanded from macro 'WSA_CMSG_SPACE'
  189 | #define WSA_CMSG_SPACE(length) (WSA_CMSGDATA_ALIGN(sizeof(WSACMSGHDR) + WSA_CMSGHDR_ALIGN(length)))
      |                                                                         ^~~~~~~~~~~~~~~~~~~~~~~~~
D:/msys64/clang64/include/mswsock.h:184:83: note: expanded from macro 'WSA_CMSGHDR_ALIGN'
  184 | #define WSA_CMSGHDR_ALIGN(length) (((length) + TYPE_ALIGNMENT(WSACMSGHDR)-1) & (~(TYPE_ALIGNMENT(WSACMSGHDR)-1)))
      |                                                                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~
D:/msys64/clang64/include/winnt.h:136:40: note: expanded from macro 'TYPE_ALIGNMENT'
  136 | #define TYPE_ALIGNMENT(t) FIELD_OFFSET(struct { char x; t test; }, test)
      |                                        ^~~~~~
D:/msys64/clang64/include/winnt.h:687:62: note: expanded from macro 'FIELD_OFFSET'
  687 | #define FIELD_OFFSET(Type, Field) ((LONG) __builtin_offsetof(Type, Field))
      |                                                              ^~~~
D:/msys64/clang64/include/mswsock.h:185:39: note: expanded from macro 'WSA_CMSGDATA_ALIGN'
  185 | #define WSA_CMSGDATA_ALIGN(length) (((length) + MAX_NATURAL_ALIGNMENT-1) & (~(MAX_NATURAL_ALIGNMENT-1)))
      |                                       ^~~~~~
D:/dev/ossia/score/3rdparty/libossia/3rdparty/libcoap/src/coap_io.c:1169:54: warning: defining a type within '__builtin_offsetof' is a C23 extension [-Wc23-extensions]
 1169 |       for (cmsg = CMSG_FIRSTHDR(&mhdr); cmsg; cmsg = CMSG_NXTHDR(&mhdr, cmsg)) {
      |                                                      ^~~~~~~~~~~~~~~~~~~~~~~~
D:/dev/ossia/score/3rdparty/libossia/3rdparty/libcoap/src/coap_io.c:758:21: note: expanded from macro 'CMSG_NXTHDR'
  758 | #define CMSG_NXTHDR WSA_CMSG_NXTHDR
      |                     ^
D:/msys64/clang64/include/mswsock.h:187:95: note: expanded from macro 'WSA_CMSG_NXTHDR'
  187 | #define WSA_CMSG_NXTHDR(msg,cmsg) ((!(cmsg)) ? WSA_CMSG_FIRSTHDR(msg) : ((((u_char *)(cmsg) + WSA_CMSGHDR_ALIGN((cmsg)->cmsg_len) + sizeof(WSACMSGHDR)) > (u_char *)((msg)->Control.buf) + (msg)->Control.len) ? (LPWSACMSGHDR)NULL : (LPWSACMSGHDR)((u_char *)(cmsg) + WSA_CMSGHDR_ALIGN((cmsg)->cmsg_len))))
      |                                                                                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
D:/msys64/clang64/include/mswsock.h:184:48: note: expanded from macro 'WSA_CMSGHDR_ALIGN'
  184 | #define WSA_CMSGHDR_ALIGN(length) (((length) + TYPE_ALIGNMENT(WSACMSGHDR)-1) & (~(TYPE_ALIGNMENT(WSACMSGHDR)-1)))
      |                                                ^~~~~~~~~~~~~~~~~~~~~~~~~~
D:/msys64/clang64/include/winnt.h:136:40: note: expanded from macro 'TYPE_ALIGNMENT'
  136 | #define TYPE_ALIGNMENT(t) FIELD_OFFSET(struct { char x; t test; }, test)
      |                                        ^~~~~~
D:/msys64/clang64/include/winnt.h:687:62: note: expanded from macro 'FIELD_OFFSET'
  687 | #define FIELD_OFFSET(Type, Field) ((LONG) __builtin_offsetof(Type, Field))
      |                                                              ^~~~
D:/dev/ossia/score/3rdparty/libossia/3rdparty/libcoap/src/coap_io.c:1169:54: warning: defining a type within '__builtin_offsetof' is a C23 extension [-Wc23-extensions]
 1169 |       for (cmsg = CMSG_FIRSTHDR(&mhdr); cmsg; cmsg = CMSG_NXTHDR(&mhdr, cmsg)) {
      |                                                      ^~~~~~~~~~~~~~~~~~~~~~~~
D:/dev/ossia/score/3rdparty/libossia/3rdparty/libcoap/src/coap_io.c:758:21: note: expanded from macro 'CMSG_NXTHDR'
  758 | #define CMSG_NXTHDR WSA_CMSG_NXTHDR
      |                     ^
D:/msys64/clang64/include/mswsock.h:187:95: note: expanded from macro 'WSA_CMSG_NXTHDR'
  187 | #define WSA_CMSG_NXTHDR(msg,cmsg) ((!(cmsg)) ? WSA_CMSG_FIRSTHDR(msg) : ((((u_char *)(cmsg) + WSA_CMSGHDR_ALIGN((cmsg)->cmsg_len) + sizeof(WSACMSGHDR)) > (u_char *)((msg)->Control.buf) + (msg)->Control.len) ? (LPWSACMSGHDR)NULL : (LPWSACMSGHDR)((u_char *)(cmsg) + WSA_CMSGHDR_ALIGN((cmsg)->cmsg_len))))
      |                                                                                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
D:/msys64/clang64/include/mswsock.h:184:83: note: expanded from macro 'WSA_CMSGHDR_ALIGN'
  184 | #define WSA_CMSGHDR_ALIGN(length) (((length) + TYPE_ALIGNMENT(WSACMSGHDR)-1) & (~(TYPE_ALIGNMENT(WSACMSGHDR)-1)))
      |                                                                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~
D:/msys64/clang64/include/winnt.h:136:40: note: expanded from macro 'TYPE_ALIGNMENT'
  136 | #define TYPE_ALIGNMENT(t) FIELD_OFFSET(struct { char x; t test; }, test)
      |                                        ^~~~~~
D:/msys64/clang64/include/winnt.h:687:62: note: expanded from macro 'FIELD_OFFSET'
  687 | #define FIELD_OFFSET(Type, Field) ((LONG) __builtin_offsetof(Type, Field))
      |                                                              ^~~~
D:/dev/ossia/score/3rdparty/libossia/3rdparty/libcoap/src/coap_io.c:1169:54: warning: defining a type within '__builtin_offsetof' is a C23 extension [-Wc23-extensions]
 1169 |       for (cmsg = CMSG_FIRSTHDR(&mhdr); cmsg; cmsg = CMSG_NXTHDR(&mhdr, cmsg)) {
      |                                                      ^~~~~~~~~~~~~~~~~~~~~~~~
D:/dev/ossia/score/3rdparty/libossia/3rdparty/libcoap/src/coap_io.c:758:21: note: expanded from macro 'CMSG_NXTHDR'
  758 | #define CMSG_NXTHDR WSA_CMSG_NXTHDR
      |                     ^
D:/msys64/clang64/include/mswsock.h:187:265: note: expanded from macro 'WSA_CMSG_NXTHDR'
  187 | #define WSA_CMSG_NXTHDR(msg,cmsg) ((!(cmsg)) ? WSA_CMSG_FIRSTHDR(msg) : ((((u_char *)(cmsg) + WSA_CMSGHDR_ALIGN((cmsg)->cmsg_len) + sizeof(WSACMSGHDR)) > (u_char *)((msg)->Control.buf) + (msg)->Control.len) ? (LPWSACMSGHDR)NULL : (LPWSACMSGHDR)((u_char *)(cmsg) + WSA_CMSGHDR_ALIGN((cmsg)->cmsg_len))))
      |                                                                                                                                                                                                                                                                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
D:/msys64/clang64/include/mswsock.h:184:48: note: expanded from macro 'WSA_CMSGHDR_ALIGN'
  184 | #define WSA_CMSGHDR_ALIGN(length) (((length) + TYPE_ALIGNMENT(WSACMSGHDR)-1) & (~(TYPE_ALIGNMENT(WSACMSGHDR)-1)))
      |                                                ^~~~~~~~~~~~~~~~~~~~~~~~~~
D:/msys64/clang64/include/winnt.h:136:40: note: expanded from macro 'TYPE_ALIGNMENT'
  136 | #define TYPE_ALIGNMENT(t) FIELD_OFFSET(struct { char x; t test; }, test)
      |                                        ^~~~~~
D:/msys64/clang64/include/winnt.h:687:62: note: expanded from macro 'FIELD_OFFSET'
  687 | #define FIELD_OFFSET(Type, Field) ((LONG) __builtin_offsetof(Type, Field))
      |                                                              ^~~~
D:/dev/ossia/score/3rdparty/libossia/3rdparty/libcoap/src/coap_io.c:1169:54: warning: defining a type within '__builtin_offsetof' is a C23 extension [-Wc23-extensions]
 1169 |       for (cmsg = CMSG_FIRSTHDR(&mhdr); cmsg; cmsg = CMSG_NXTHDR(&mhdr, cmsg)) {
      |                                                      ^~~~~~~~~~~~~~~~~~~~~~~~
D:/dev/ossia/score/3rdparty/libossia/3rdparty/libcoap/src/coap_io.c:758:21: note: expanded from macro 'CMSG_NXTHDR'
  758 | #define CMSG_NXTHDR WSA_CMSG_NXTHDR
      |                     ^
D:/msys64/clang64/include/mswsock.h:187:265: note: expanded from macro 'WSA_CMSG_NXTHDR'
  187 | #define WSA_CMSG_NXTHDR(msg,cmsg) ((!(cmsg)) ? WSA_CMSG_FIRSTHDR(msg) : ((((u_char *)(cmsg) + WSA_CMSGHDR_ALIGN((cmsg)->cmsg_len) + sizeof(WSACMSGHDR)) > (u_char *)((msg)->Control.buf) + (msg)->Control.len) ? (LPWSACMSGHDR)NULL : (LPWSACMSGHDR)((u_char *)(cmsg) + WSA_CMSGHDR_ALIGN((cmsg)->cmsg_len))))
      |                                                                                                                                                                                                                                                                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
D:/msys64/clang64/include/mswsock.h:184:83: note: expanded from macro 'WSA_CMSGHDR_ALIGN'
  184 | #define WSA_CMSGHDR_ALIGN(length) (((length) + TYPE_ALIGNMENT(WSACMSGHDR)-1) & (~(TYPE_ALIGNMENT(WSACMSGHDR)-1)))
      |                                                                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~
D:/msys64/clang64/include/winnt.h:136:40: note: expanded from macro 'TYPE_ALIGNMENT'
  136 | #define TYPE_ALIGNMENT(t) FIELD_OFFSET(struct { char x; t test; }, test)
      |                                        ^~~~~~
D:/msys64/clang64/include/winnt.h:687:62: note: expanded from macro 'FIELD_OFFSET'
  687 | #define FIELD_OFFSET(Type, Field) ((LONG) __builtin_offsetof(Type, Field))
      |                                                              ^~~~
14 warnings generated.
```

mrdeep1 commented:
Thanks for confirming it generally works.

When testing CLANG64, I had added `-DCMAKE_C_FLAGS=-Wno-c23-extensions` to get rid of these complaints.  I have just added this option into CMakeLists.txt and pushed the code change.

mrdeep1 commented:
@jcelerier Does all now work cleanly for you?

jcelerier commented:
yes, thanks!
